ideogram-4 fp8 NotImplementedError: "mul_cuda" not implemented for 'Float8_e4m3fn'

### Custom Node Testing

- [x] I have tried disabling custom nodes and the issue persists (see [how to disable custom nodes](https://docs.comfy.org/troubleshooting/custom-node-issues#step-1%3A-test-with-all-custom-nodes-disabled) if you need help)

### Expected Behavior

Loading and running the Ideogram 4 checkpoint (which is only available in fp8_e4m3fn format) crashes on the first sampling step with:
    
    
    NotImplementedError: "mul_cuda" not implemented for 'Float8_e4m3fn'
    
    
    The traceback terminates at comfy/ldm/ideogram4/model.py, line 161, in _backbone:
    python
    x = x * output_image_mask
    
    
    The model stores self.dtype = torch.float8_e4m3fn and _img_to_tokens() casts image tokens to that dtype. Then _run_conditional() builds x_full as torch.zeros(B, L, latent_dim, dtype=img_tokens.dtype, ...) -- which is also fp8. When _backbone calls x * output_image_mask, PyTorch refuses because element-wise mul is not implemented for fp8 on CUDA. Only matmul via scaled_mm/comfy_kitchen is supported.

### Actual Behavior

    File "comfy/ldm/ideogram4/model.py", line 297, in _forward
        return -self._run_conditional(x, context, attention_mask, timesteps, gh, gw, transformer_options)
    File "comfy/ldm/ideogram4/model.py", line 264, in _run_conditional
        out = self._backbone(context_chunk, x_full, t_chunk, position_ids, attn_mask, indicator, ...)
    File "comfy/ldm/ideogram4/model.py", line 161, in _backbone
        x = x * output_image_mask
    NotImplementedError: "mul_cuda" not implemented for 'Float8_e4m3fn'


### Steps to Reproduce

    1. Download the Ideogram 4 fp8 checkpoint
    2. Load via CheckpointLoaderAdvanced or SimpleCheckpointLoader
    3. Use any standard sampler workflow (SamplerCustomAdvanced, KSampler, etc.)
    4. Queue prompt -- crashes on first step


### Debug Logs

```powershell
[ERROR] !!! Exception during processing !!! "mul_cuda" not implemented for 'Float8_e4m3fn'
[ERROR] Traceback (most recent call last):
  File "/home/anon/Documents/ComfyUI/execution.py", line 536, in execute
    output_data, output_ui, has_subgraph, has_pending_tasks = await get_output_data(prompt_id, unique_id, obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, v3_data=v3_data)
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/execution.py", line 336, in get_output_data
    return_values = await _async_map_node_over_list(prompt_id, unique_id, obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, v3_data=v3_data)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/execution.py", line 310, in _async_map_node_over_list
    await process_inputs(input_dict, i)
  File "/home/anon/Documents/ComfyUI/execution.py", line 298, in process_inputs
    result = f(**inputs)
  File "/home/anon/Documents/ComfyUI/comfy_api/internal/__init__.py", line 149, in wrapped_func
    return method(locked_class, **inputs)
  File "/home/anon/Documents/ComfyUI/comfy_api/latest/_io.py", line 1880, in EXECUTE_NORMALIZED
    to_return = cls.execute(*args, **kwargs)
  File "/home/anon/Documents/ComfyUI/comfy_extras/nodes_custom_sampler.py", line 1047, in execute
    samples = guider.sample(noise.generate_noise(latent), latent_image, sampler, sigmas, denoise_mask=noise_mask, callback=callback, disable_pbar=disable_pbar, seed=noise.seed)
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 1316, in sample
    output = executor.execute(noise, latent_image, sampler, sigmas, denoise_mask, callback, disable_pbar, seed, latent_shapes=latent_shapes)
  File "/home/anon/Documents/ComfyUI/comfy/patcher_extension.py", line 113, in execute
    return self.original(*args, **kwargs)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy_extras/nodes_custom_sampler.py", line 918, in outer_sample
    return super().outer_sample(noise, latent_image, sampler, sigmas, denoise_mask, callback, disable_pbar, seed, latent_shapes=latent_shapes)
           ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 1254, in outer_sample
    output = self.inner_sample(noise, latent_image, device, sampler, sigmas, denoise_mask, callback, disable_pbar, seed, latent_shapes=latent_shapes)
  File "/home/anon/Documents/ComfyUI/comfy_extras/nodes_custom_sampler.py", line 932, in inner_sample
    return super().inner_sample(noise, latent_image, device, sampler, sigmas, denoise_mask, callback, disable_pbar, seed, latent_shapes=latent_shapes)
           ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 1229, in inner_sample
    samples = executor.execute(self, sigmas, extra_args, callback, noise, latent_image, denoise_mask, disable_pbar)
  File "/home/anon/Documents/ComfyUI/comfy/patcher_extension.py", line 113, in execute
    return self.original(*args, **kwargs)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 999, in sample
    samples = self.sampler_function(model_k, noise, sigmas, extra_args=extra_args, callback=k_callback, disable=disable_pbar, **self.extra_options)
  File "/home/anon/miniconda3/envs/comfy/lib/python3.13/site-packages/torch/utils/_contextlib.py", line 124, in decorate_context
    return func(*args, **kwargs)
  File "/home/anon/Documents/ComfyUI/comfy/k_diffusion/sampling.py", line 1460, in sample_res_multistep
    return res_multistep(model, x, sigmas, extra_args=extra_args, callback=callback, disable=disable, s_noise=s_noise, noise_sampler=noise_sampler, eta=0., cfg_pp=False)
  File "/home/anon/miniconda3/envs/comfy/lib/python3.13/site-packages/torch/utils/_contextlib.py", line 124, in decorate_context
    return func(*args, **kwargs)
  File "/home/anon/Documents/ComfyUI/comfy/k_diffusion/sampling.py", line 1418, in res_multistep
    denoised = model(x, sigmas[i] * s_in, **extra_args)
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 639, in __call__
    out = self.inner_model(x, sigma, model_options=model_options, seed=seed)
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 1202, in __call__
    return self.outer_predict_noise(*args, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 1209, in outer_predict_noise
    ).execute(x, timestep, model_options, seed)
      ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/patcher_extension.py", line 114, in execute
    return self.wrappers[self.idx](self, *args, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy_extras/nodes_custom_sampler.py", line 1164, in predict_noise_wrapper
    return executor(*args, **kwargs)
  File "/home/anon/Documents/ComfyUI/comfy/patcher_extension.py", line 106, in __call__
    return new_executor.execute(*args, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/patcher_extension.py", line 113, in execute
    return self.original(*args, **kwargs)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy_extras/nodes_custom_sampler.py", line 938, in predict_noise
    cond = comfy.samplers.calc_cond_batch(self.inner_model, [positive], x, timestep, model_options)[0]
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 210, in calc_cond_batch
    return _calc_cond_batch_outer(model, conds, x_in, timestep, model_options)
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 218, in _calc_cond_batch_outer
    return executor.execute(model, conds, x_in, timestep, model_options)
           ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/patcher_extension.py", line 113, in execute
    return self.original(*args, **kwargs)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/samplers.py", line 334, in _calc_cond_batch
    output = model.apply_model(input_x, timestep_, **c).chunk(batch_chunks)
             ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/model_base.py", line 187, in apply_model
    return comfy.patcher_extension.WrapperExecutor.new_class_executor(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ...<2 lines>...
        comfy.patcher_extension.get_all_wrappers(comfy.patcher_extension.WrappersMP.APPLY_MODEL, transformer_options)
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ).execute(x, t, c_concat, c_crossattn, control, transformer_options, **kwargs)
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/patcher_extension.py", line 113, in execute
    return self.original(*args, **kwargs)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/model_base.py", line 231, in _apply_model
    model_output = self.diffusion_model(xc, t, context=context, control=control, transformer_options=transformer_options, **extra_conds)
  File "/home/anon/miniconda3/envs/comfy/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1778, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/anon/miniconda3/envs/comfy/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1789, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/anon/Documents/ComfyUI/comfy/ldm/ideogram4/model.py", line 282, in forward
    return comfy.patcher_extension.WrapperExecutor.new_class_executor(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ...<2 lines>...
        comfy.patcher_extension.get_all_wrappers(comfy.patcher_extension.WrappersMP.DIFFUSION_MODEL, transformer_options),
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ).execute(x, timesteps, context, attention_mask, transformer_options, **kwargs)
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/patcher_extension.py", line 113, in execute
    return self.original(*args, **kwargs)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/ldm/ideogram4/model.py", line 297, in _forward
    return -self._run_conditional(x, context, attention_mask, timesteps, gh, gw, transformer_options)
            ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anon/Documents/ComfyUI/comfy/ldm/ideogram4/model.py", line 264, in _run_conditional
    out = self._backbone(context_chunk, x_full, t_chunk, position_ids, attn_mask, indicator,
                         transformer_options=transformer_options)
  File "/home/anon/Documents/ComfyUI/comfy/ldm/ideogram4/model.py", line 161, in _backbone
    x = x * output_image_mask
        ~~^~~~~~~~~~~~~~~~~~~
NotImplementedError: "mul_cuda" not implemented for 'Float8_e4m3fn'
```

### Other

    System Information
    - ComfyUI Version: 0.24.0
    - PyTorch Version: 2.12.0+cu130
    - GPU: NVIDIA GeForce RTX 4090 D (48GB) + RTX 3090 (24GB)
    - OS: Linux
    - Python Version: 3.13.11

--- a/comfy/ldm/ideogram4/model.py
+++ b/comfy/ldm/ideogram4/model.py
@@ -174,7 +174,7 @@ class Ideogram4Transformer(nn.Module):
             llm = self.llm_cond_proj(llm) * text_mask
             h[:, :L_text] = h[:, :L_text] + llm
 
-        h = h + self.embed_image_indicator((indicator == OUTPUT_IMAGE_INDICATOR).to(torch.long))
+        h = h + self.embed_image_indicator((indicator == OUTPUT_IMAGE_INDICATOR).to(torch.long), out_dtype=h.dtype)
 
         # Qwen3-VL interleaved MRoPE; position_ids (B, L, 3) -> (3, L) (same across batch).
         freqs_cis = precompute_freqs_cis(
@@ -235,7 +235,7 @@ class Ideogram4Transformer2DModel(Ideogram4Transformer):
     def _run_conditional(self, x_chunk, context_chunk, attn_mask_chunk, t_chunk, gh, gw, transformer_options):
         B = x_chunk.shape[0]
         device = x_chunk.device
-        img_tokens = self._img_to_tokens(x_chunk).to(self.dtype)
+        img_tokens = self._img_to_tokens(x_chunk)
         L_img = img_tokens.shape[1]
         L_text = context_chunk.shape[1]
         L = L_text + L_img
@@ -268,7 +268,7 @@ class Ideogram4Transformer2DModel(Ideogram4Transformer):
     def _run_image_only(self, x_chunk, t_chunk, gh, gw, transformer_options):
         B = x_chunk.shape[0]
         device = x_chunk.device
-        img_tokens = self._img_to_tokens(x_chunk).to(self.dtype)
+        img_tokens = self._img_to_tokens(x_chunk)
         L_img = img_tokens.shape[1]
 
         position_ids = self._image_position_ids(gh, gw, device).unsqueeze(0).expand(B, L_img, 3)

